### PR TITLE
Print all test case error messages together at the end

### LINKED ISSUE #744: When tests fail, it is hard to tell which test it was
When tests fail, we get a very long string of output in GitHub Actions that has much important information, but missing something that I quite care about: What the name of the test actually is! (not it's hash/slug, but the regular name).

Below is an example:
```console
   mfc: OK > (venv) Entered the Python 3.11.5 virtual environment (>= 3.9).

      .=++*:          -+*+=.        | sbryngelson@frontier10205 [Linux]
     :+   -*-        ==   =* .      | ---------------------------------
   :*+      ==      ++    .+-       | 
  :*##-.....:*+   .#%+++=--+=:::.   | --jobs 1
  -=-++-======#=--**+++==+*++=::-:. | --mpi --gpu --no-debug --no-gcov --no-unified
 .:++=----------====+*= ==..:%..... | --targets pre_process and simulation
  .:-=++++===--==+=-+=   +.  :=     | 
  +#=::::::::=%=. -+:    =+   *:    | ----------------------------------------------------------
 .*=-=*=..    :=+*+:      -...--    | $ ./mfc.sh (build, run, test, clean, count, packer) --help

 Acquiring /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/tests/9EB947DB/case.py...
 Acquiring /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/tests/9EB947DB/case.py...
 Build | syscheck, syscheck, pre_process, and simulation | Generic Build


 Run
   Using queue system Interactive.
   Using baked-in template for frontier.
   Generating input files for syscheck...
    
     Generating syscheck.inp:
       INFO: Forwarded 0/95 parameters.
    
   Generating input files for pre_process...
    
     Generating pre_process.inp:
       INFO: Forwarded 69/95 parameters.
    
   Generating input files for simulation...
    
     Generating simulation.inp:
       INFO: Forwarded 59/95 parameters.
    
   $ /bin/bash /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/tests/9EB947DB/MFC.sh

+-----------------------------------------------------------------------------------------------------------+
| MFC case # MFC @ /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/tests/9EB947DB/case.py: |
+-----------------------------------------------------------------------------------------------------------+
| * Start-time     11:42:41                            * Start-date     11:42:41                            |
| * Partition      N/A                                 * Walltime       01:00:00                            |
| * Account        N/A                                 * Nodes          1                                   |
| * Job Name       MFC                                 * Engine         interactive                         |
| * QoS            N/A                                 * Binary         N/A                                 |
| * Queue System   Interactive                         * Email          N/A                                 |
+-----------------------------------------------------------------------------------------------------------+

mfc: OK > :) Loading modules:


mfc: OK > :) Running syscheck:

+ srun --nodes 1 --ntasks-per-node 1 --cpus-per-task 7 --gpus-per-task 1 --gpu-bind closest /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/build/install/ab29bd3004/bin/syscheck
 [TEST] MPI: call mpi_init(ierr)
 [TEST] MPI: call mpi_comm_rank(MPI_COMM_WORLD, rank, ierr)
 [TEST] MPI: call mpi_barrier(MPI_COMM_WORLD, ierr)
 [TEST] MPI: call assert(rank >= 0)
 [TEST] MPI: call mpi_comm_size(MPI_COMM_WORLD, nRanks, ierr)
 [TEST] MPI: call assert(nRanks > 0 .and. rank < nRanks)
 [TEST] ACC: devtype = acc_get_device_type()
 [TEST] ACC: num_devices = acc_get_num_devices(devtype)
 [TEST] ACC: call assert(num_devices > 0)
 [TEST] ACC: call acc_set_device_num(mod(rank, nRanks), devtype)
 [TEST] ACC: allocate(arr(1:N))
 [TEST] ACC: !$acc enter data create(arr(1:N))
 [TEST] ACC: !$acc parallel loop
 [TEST] ACC: !$acc update host(arr(1:N))
 [TEST] ACC: !$acc exit data delete(arr)
 [TEST] MPI: call mpi_barrier(MPI_COMM_WORLD, ierr)
 [TEST] MPI: call mpi_finalize(ierr)
 
 Syscheck: PASSED.

mfc: OK > :) Running pre_process:

+ srun --nodes 1 --ntasks-per-node 1 --cpus-per-task 7 --gpus-per-task 1 --gpu-bind closest /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/build/install/4b354ae468/bin/pre_process
 Pre-processing a 299x0x0 case on 1 rank(s)
 Processing patch 1
 Processing patch 2
 Processing patch 3
 Elapsed Time 1.62700000000000344E-3

mfc: OK > :) Running simulation:

+ srun --nodes 1 --ntasks-per-node 1 --cpus-per-task 7 --gpus-per-task 1 --gpu-bind closest /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/build/install/f1b760553b/bin/simulation
ACC:            find_in_present_table failed for 'fd_coeff_x(:,:)' (0x2ebd0c0-0x2ebffa0) from ../../../lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/src/simulation/m_hypoelastic.fpp:75
ACC: libcrayacc/acc_runtime.c:877 CRAY_ACC_ERROR - Variable not found in present table
 Simulating a regular 299x0x0 case on 1 rank(s) with OpenACC offloading.
srun: error: frontier10205: task 0: Exited with exit code 1
srun: Terminating StepId=2788488.245

mfc: ERROR > :( /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/build/install/f1b760553b/bin/simulation failed with exit code 1.

 

Error: Submitting batch file for Interactive failed. It can be found here: /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/tests/9EB947DB/MFC.sh. Please check the file for errors.

./mfc.sh: line 53: 659792 Terminated              python3 "$(pwd)/toolchain/main.py" "$@"

mfc: ERROR > main.py finished with a 143 exit code.
mfc: (venv) Exiting the Python virtual environment.

    Attempt 1: Failed test 9EB947DB. Retrying...
   mfc: OK > (venv) Entered the Python 3.11.5 virtual environment (>= 3.9).

      .=++*:          -+*+=.        | sbryngelson@frontier10205 [Linux]
     :+   -*-        ==   =* .      | ---------------------------------
   :*+      ==      ++    .+-       | 
  :*##-.....:*+   .#%+++=--+=:::.   | --jobs 1
  -=-++-======#=--**+++==+*++=::-:. | --mpi --gpu --no-debug --no-gcov --no-unified
 .:++=----------====+*= ==..:%..... | --targets pre_process and simulation
  .:-=++++===--==+=-+=   +.  :=     | 
  +#=::::::::=%=. -+:    =+   *:    | ----------------------------------------------------------
 .*=-=*=..    :=+*+:      -...--    | $ ./mfc.sh (build, run, test, clean, count, packer) --help

 Acquiring /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/tests/AF0BCEE4/case.py...
 Acquiring /lustre/orion/cfd154/scratch/sbryngelson/runner/actions-runner/_work/MFC/MFC/tests/AF0BCEE4/case.py...
 Build | syscheck, syscheck, pre_process, and simulation | Generic Build
```

As you can see, this makes it quite hard to diagnose what cases are going wrong.

I suggest two rather straightforward things to resolve this GH Issue and make this headache go away.

1. Whenever it runs a test, make the stdout also print the test case name along with the hash. There are multiple routes to success here. I'm not sure what is easiest.
2. At the end of the test suite, it says, `X jobs passed, Y failed, Z skipped`. This should really list the jobs that failed and the jobs that were skipped (but not passed; that's going to be too many).

### PR DESCRIPTION
**Pull request by wilfonba**
## Description

I have recently found the behavior described in #744 to be annoying. This PR addresses that by keeping a list of all failed tests and printing the errors all together at the end.

Fixes #744

### Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

### Scope

- [x] This PR comprises a set of related changes with a common goal

## How Has This Been Tested?

Test 1: I intentionally caused tests to fail due to tolerance by commenting out `dvd(0)` in the fifth order WENO reconstruction so the solution would be wrong. The resulting output of `./mfc.sh test -j 1 -o bc=-1 > test1.txt` (without coloring) is
[test1.txt](https://github.com/user-attachments/files/17891517/test1.txt)

Test 2: Intentionally caused tests to fail due to NaNs by commenting out `beta(0)` in the fifth-order WENO reconstruction. The resulting output of `./mfc.sh test -j 1 -o bc=-1 > test2.txt` (without color) is 
[test2.txt](https://github.com/user-attachments/files/17891516/test2.txt)

Test 3: Intentionally caused tolerance errors in the WENO routine and skipped some tests. The result of running `./mfc.sh test -j 8 -o 1D -% 80 > test4.txt` (without color) is 
[test4.txt](https://github.com/user-attachments/files/17891633/test4.txt)

### CODE REVIEW

File: test4.txt
Comment (sbryngelson) [inline test4.txt:1]: remove this file (obviously)

**Review by sbryngelson — CHANGES_REQUESTED** [review]
@henryleberre might comment on the 'cleanness' of the toolchain implementation. For me, it seems reasonable and gets the job done.

**Comment by sbryngelson** [comment]
Is this in conflict with something added by @henryleberre in https://github.com/MFlowCode/MFC/pull/737 ?

**Comment by wilfonba** [comment]
Merged without issue

**Comment by sbryngelson** [comment]
@wilfonba https://github.com/MFlowCode/MFC/pull/737 is still open (not merged)

**Review by sbryngelson — APPROVED** [review]

**Pull request merged**